# Install packages

In [ ]:
!pip3 install ffmpeg-python deepspeed

# Import libraries

In [ ]:
from google.colab import runtime
runtime.unassign()

In [ ]:
import os
import shutil
import ffmpeg
import tarfile
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')


Mounted at /content/gdrive


# Dataset Preparation

In [ ]:
import pandas as pd
df = pd.read_excel('/content/gdrive/MyDrive/85cleaned_file.xlsx')
# df1 = pd.read_csv('/content/gdrive/MyDrive/finetuning_dataset/audio_text_youssef.csv')
# df2 = pd.read_csv('/content/gdrive/MyDrive/finetuning_dataset/audio_text_amir.csv')

# df = pd.concat([df1, df2])

In [ ]:
# df = df[df['is_remove'] == 0]
df

,Chunk Name,Start Time (ms),End Time (ms),Duration (s),Chunk Path,Transcript,Word Count,Words Per Second,Is_Music,Detected_Emotions,Is_Denoised,phonemes,transcript_with_emotions,PlayAudio,Unnamed: 14,Unnamed: 15
0,audio001-chunk0.wav,0,13124,13.124,dataset\audio001-chunk0.wav,خليني نقول انا ابن عائله جدا متواضعه طيب وين ت...,30,2.285888,False,NaN,0,knɪŋ gʊ ənə ɪ b n tæ ɪlɪʃɪ dɪnm təwəðiaɪ fɪʃ ...,خليني نقول انا ابن عائله جدا متواضعه طيب وين ت...,▶ Play,NaN,NaN
1,audio001-chunk1.wav,13757,36498,22.741,dataset\audio001-chunk1.wav,عامه يقرا كتاب يحس روحه باللي را ناجح انت غير ...,31,1.363177,False,"{'Laughter': {'confidence': 0.142, 'start_time...",0,khɪ kənə k tə jihəsəhə b lɪhnɛʃ təɪ kəni tə k...,عامه يقرا كتاب <Laughter> يحس روحه باللي را نا...,▶ Play,باللي را ناجح انت غير قيت كتا,NaN
2,audio001-chunk2.wav,37027,42197,5.170,dataset\audio001-chunk2.wav,في حياتهم يا اما فر يا اماي,7,1.353965,False,NaN,0,jɛ ɛməðeɪɑhaɪ ə ʧivɝz jɛ ɛməðeɪɑɹəloʊ dɛɹ ən d...,في حياتهم يا اما فر يا اماي,▶ Play,NaN,NaN
3,audio001-chunk3.wav,42386,47554,5.168,dataset\audio001-chunk3.wav,اماي اليده وهرانيه جبتلكم بني ادم فها قسره ضحك,9,1.741486,False,NaN,0,hʊ ziumə gə dəwhɑɾɑnioʊ ʃɪv t kʊm zuʃ b niɛ də...,اماي اليده وهرانيه جبتلكم بني ادم فها قسره ضحك,▶ Play,NaN,NaN
4,audio001-chunk4.wav,47828,51925,4.097,dataset\audio001-chunk4.wav,تعرفت عليهم هنا في دباي ومن مدينتي من,8,1.952648,False,NaN,0,ə gʊsənɑ dɑh k tɾəf tɛɾioʊmhənəfɪ du baɪ wumɪm...,تعرفت عليهم هنا في دباي ومن مدينتي من,▶ Play,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8156,audio022-chunk339.wav,2320814,2326703,5.889,dataset\audio022-chunk339.wav,كي تخرج لك هنايا طاك واش تروح دير استنى هذا يط...,17,2.886738,False,NaN,0,kɪ tfɑʃɪ kənæɑ tɑ k wɪʃ t əhɪ dɪɪ sθɪnəhə deɪ...,كي تخرج لك هنايا طاك واش تروح دير استنى هذا يط...,▶ Play,NaN,NaN
8157,audio022-chunk340.wav,2326832,2330809,3.977,dataset\audio022-chunk340.wav,كييقوللك اوكي طاك واش ندير بابليش,6,1.508675,False,NaN,0,ki oʊlɪ k oʊ keɪtɑ k wuʃn dɪɾoʊ bɑ b leɪ,كييقوللك اوكي طاك واش ندير بابليش,▶ Play,NaN,NaN
8158,audio022-chunk341.wav,2331038,2334331,3.293,dataset\audio022-chunk341.wav,بابليش كليكي على هذيا من بعد تخرجلك لابيبليس تاعك,9,2.733070,False,NaN,0,ʃ k li k ɛɾəhə di mən bæ dɪ təhɑɹ ʤlɪ k lə pɪ ...,بابليش كليكي على هذيا من بعد تخرجلك لابيبليس تاعك,▶ Play,NaN,NaN
8159,audio022-chunk342.wav,2334605,2354064,19.459,dataset\audio022-chunk342.wav,تاعك هنايا واجده هنا في الديليفري يكتبلك باللي...,50,2.569505,False,NaN,0,hnæɑ wɛʃ daʊ hnɛf diɪvɪɾi ɪ tɪ b lɪ bɪi jɛni...,تاعك هنايا واجده هنا في الديليفري يكتبلك باللي...,▶ Play,NaN,NaN


## Remove audios

In [ ]:
# audios_path = '/content/gdrive/MyDrive/finetuning_dataset/dataset/audios'
# audios = os.listdir(audios_path)

# for audio in audios:
#   current_audio_path = os.path.join(audios_path, audio)
#   os.remove(current_audio_path)

## Copy audios to another folder

In [ ]:

# os.makedirs('/content/gdrive/MyDrive/finetuning_dataset/dataset/wavs', exist_ok=True)

# source_path = '/content/gdrive/MyDrive/finetuning_dataset/dataset/splitted_audios'
# distination_path = '/content/gdrive/MyDrive/finetuning_dataset/dataset/wavs'

# audios = df['filename'].tolist()
# for audio in audios:
#   if audio.endswith('.wav'):
#     current_audio_path = os.path.join(source_path, audio)
#     new_audio_path = os.path.join(distination_path, audio)
#     shutil.copy2(current_audio_path, new_audio_path)

In [ ]:
df.drop(columns=['is_remove', 'duration'], axis=1, inplace=True)

In [ ]:
df['filename'] = df['filename'].apply(lambda x: x.replace('.wav', ''))
# df['text'] = df['text'].apply(lambda x: str(x)+'|')
# df['normalized_text'] = df['normalized_text'].apply(lambda x: str(x)+'|')

In [ ]:
import os
import pandas as pd


# Create the new DataFrame with required columns
df2 = pd.DataFrame({
    "filename": df["Chunk Name"],
    "text": df["Transcript"],
    "normalized_text": df["Transcript"]  # Just copying for now
})




In [ ]:
df2.to_csv('/content/gdrive/MyDrive/dataset22050/metadata.csv', encoding='utf-8-sig', index=False, header=False, sep='|',)

In [ ]:
df2


,filename,text,normalized_text
0,audio001-chunk0.wav,خليني نقول انا ابن عائله جدا متواضعه طيب وين ت...,خليني نقول انا ابن عائله جدا متواضعه طيب وين ت...
1,audio001-chunk1.wav,عامه يقرا كتاب يحس روحه باللي را ناجح انت غير ...,عامه يقرا كتاب يحس روحه باللي را ناجح انت غير ...
2,audio001-chunk2.wav,في حياتهم يا اما فر يا اماي,في حياتهم يا اما فر يا اماي
3,audio001-chunk3.wav,اماي اليده وهرانيه جبتلكم بني ادم فها قسره ضحك,اماي اليده وهرانيه جبتلكم بني ادم فها قسره ضحك
4,audio001-chunk4.wav,تعرفت عليهم هنا في دباي ومن مدينتي من,تعرفت عليهم هنا في دباي ومن مدينتي من
...,...,...,...
8156,audio022-chunk339.wav,كي تخرج لك هنايا طاك واش تروح دير استنى هذا يط...,كي تخرج لك هنايا طاك واش تروح دير استنى هذا يط...
8157,audio022-chunk340.wav,كييقوللك اوكي طاك واش ندير بابليش,كييقوللك اوكي طاك واش ندير بابليش
8158,audio022-chunk341.wav,بابليش كليكي على هذيا من بعد تخرجلك لابيبليس تاعك,بابليش كليكي على هذيا من بعد تخرجلك لابيبليس تاعك
8159,audio022-chunk342.wav,تاعك هنايا واجده هنا في الديليفري يكتبلك باللي...,تاعك هنايا واجده هنا في الديليفري يكتبلك باللي...


In [ ]:
metadata_path = os.path.join('/content/gdrive/MyDrive/dataset22050', "metadata.csv")
with open(metadata_path, "r", encoding="utf-8") as f:
    for i, line in enumerate(f, 1):
      if i > 5:
        break
      cols = line.strip().split("|")
      if len(cols) < 3:  # Check if there are fewer than 3 columns
          print(f"Error on line {i}: {line}")
      else:
          print(f"Line {i} is valid: {cols[2]}")

Line 1 is valid: خليني نقول انا ابن عائله جدا متواضعه طيب وين تقدر تقول همي الوحيد كان نخدم كنت نبيع قصبر كبيع بورسات نخدم باش نجاوب على الغلطه اللي يطيح فيها الجزائري بصفه
Line 2 is valid: عامه يقرا كتاب يحس روحه باللي را ناجح انت غير قيت كتا كيتفرج كيتفرج على اندر اندر ناجح تتفرج على واح ناجح ما تنج بالعكس تلا واحدي دوها عندهم ترا في
Line 3 is valid: في حياتهم يا اما فر يا اماي
Line 4 is valid: اماي اليده وهرانيه جبتلكم بني ادم فها قسره ضحك
Line 5 is valid: تعرفت عليهم هنا في دباي ومن مدينتي من


In [ ]:
len(os.listdir('/content/gdrive/MyDrive/dataset'))

8191

## Change audio sample rate

In [ ]:
import time

def convert_sample_rate(audios_path, output_path, target_sample_rate):
  os.makedirs(output_path, exist_ok=True)
  audios = os.listdir(audios_path)
  start = time.time()
  for i, audio in enumerate(audios):
    if i % 100 == 0:
      current = time.time()
      print(f'Time take to process {i} audios: {current-start:.1f} seconds')
    audio_path = os.path.join(audios_path, audio)
    audio_output_path = os.path.join(output_path, audio)
    try:
      ffmpeg.input(audio_path).output(audio_output_path, ar=target_sample_rate).run(capture_stdout=True, capture_stderr=True) # .overwrite_output()
      print(f'Converted audio {i}')
      print(f"Successfully converted {audio_path} to {target_sample_rate} Hz")
    except ffmpeg.Error as e:
      print(f"Conversion failed. Error: {e.stderr.decode()}")


audios_path = '/content/gdrive/MyDrive/dataset'
output_path = '/content/gdrive/MyDrive/dataset22050'
target_sample_rate = 22050
convert_sample_rate(audios_path, output_path, target_sample_rate)

Streaming output truncated to the last 5000 lines.
Converted audio 5703
Successfully converted /content/gdrive/MyDrive/dataset/audio008-chunk377.wav to 22050 Hz
Converted audio 5704
Successfully converted /content/gdrive/MyDrive/dataset/audio008-chunk416.wav to 22050 Hz
Converted audio 5705
Successfully converted /content/gdrive/MyDrive/dataset/audio008-chunk371.wav to 22050 Hz
Converted audio 5706
Successfully converted /content/gdrive/MyDrive/dataset/audio008-chunk417.wav to 22050 Hz
Converted audio 5707
Successfully converted /content/gdrive/MyDrive/dataset/audio009-chunk4.wav to 22050 Hz
Converted audio 5708
Successfully converted /content/gdrive/MyDrive/dataset/audio009-chunk77.wav to 22050 Hz
Converted audio 5709
Successfully converted /content/gdrive/MyDrive/dataset/audio009-chunk110.wav to 22050 Hz
Converted audio 5710
Successfully converted /content/gdrive/MyDrive/dataset/audio009-chunk29.wav to 22050 Hz
Converted audio 5711
Successfully converted /content/gdrive/MyDrive/datas

# FINE-TUNING XTTS-v2 #

In [ ]:
!pip install git+https://github.com/coqui-ai/TTS

  Cloning https://github.com/coqui-ai/TTS to /tmp/pip-req-build-cl03d4_i
  Running command git clone --filter=blob:none --quiet https://github.com/coqui-ai/TTS /tmp/pip-req-build-cl03d4_i
  Resolved https://github.com/coqui-ai/TTS to commit dbf1a08a0d4e47fdad6172e433eeb34bc6b13b4e
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done


In [ ]:
!pip install transformers==4.37.1

In [ ]:
###updated training zone####

In [ ]:
from trainer import Trainer, TrainerArgs
#from trainer.logging.wandb_logger import WandbLogger
from TTS.tts.configs.shared_configs import BaseDatasetConfig
from TTS.tts.datasets import load_tts_samples
from TTS.tts.layers.xtts.trainer.gpt_trainer import GPTArgs, GPTTrainer, GPTTrainerConfig, XttsAudioConfig
from TTS.utils.manage import ModelManager

import sys
import os
import wandb

### Monkey Patching for wandb

In [ ]:
from trainer.logging.wandb_logger import WandbLogger

In [ ]:
def add_artifact(self, file_or_dir, name, artifact_type, aliases=None):
    ###instead of adding artifact, do nothing###
    print(f"========Ignoring artifact: {name} {file_or_dir}========")
    return


WandbLogger.add_artifact = add_artifact

In [ ]:
# Logging parameters
RUN_NAME = "finetuned_egtts_v2"
PROJECT_NAME = "dz_ar_model"
DASHBOARD_LOGGER = "wandb"
LOGGER_URI = None

### Dir for Training Run

In [ ]:
OUT_PATH = '/content/gdrive/MyDrive/models'
os.makedirs(OUT_PATH, exist_ok=True)

Retreive the base model files.

In [ ]:
# Define the path where XTTS v2.0.1 files will be downloaded
CHECKPOINTS_OUT_PATH = os.path.join(OUT_PATH, "EGTTS_v2.0_original_model_files/")
os.makedirs(CHECKPOINTS_OUT_PATH, exist_ok=True)

# DVAE files
DVAE_CHECKPOINT_LINK = "https://coqui.gateway.scarf.sh/hf-coqui/XTTS-v2/main/dvae.pth"
MEL_NORM_LINK = "https://coqui.gateway.scarf.sh/hf-coqui/XTTS-v2/main/mel_stats.pth"

# Set the path to the downloaded files
DVAE_CHECKPOINT = os.path.join(CHECKPOINTS_OUT_PATH, os.path.basename(DVAE_CHECKPOINT_LINK))
MEL_NORM_FILE = os.path.join(CHECKPOINTS_OUT_PATH, os.path.basename(MEL_NORM_LINK))

# download DVAE files if needed
if not os.path.isfile(DVAE_CHECKPOINT) or not os.path.isfile(MEL_NORM_FILE):
    print(" > Downloading DVAE files!")
    ModelManager._download_model_files([MEL_NORM_LINK, DVAE_CHECKPOINT_LINK], CHECKPOINTS_OUT_PATH, progress_bar=True)

# Download XTTS v2.0 checkpoint if needed
TOKENIZER_FILE_LINK = "https://coqui.gateway.scarf.sh/hf-coqui/XTTS-v2/main/vocab.json"
XTTS_CHECKPOINT_LINK = "https://coqui.gateway.scarf.sh/hf-coqui/XTTS-v2/main/model.pth"

# XTTS transfer learning parameters: You we need to provide the paths of XTTS model checkpoint that you want to do the fine tuning.
TOKENIZER_FILE = os.path.join(CHECKPOINTS_OUT_PATH, os.path.basename(TOKENIZER_FILE_LINK))  # vocab.json file
XTTS_CHECKPOINT = os.path.join(CHECKPOINTS_OUT_PATH, os.path.basename(XTTS_CHECKPOINT_LINK))  # model.pth file

# download XTTS v2.0 files if needed
if not os.path.isfile(TOKENIZER_FILE) or not os.path.isfile(XTTS_CHECKPOINT):
    print(" > Downloading XTTS v2.0 files!")
    ModelManager._download_model_files(
        [TOKENIZER_FILE_LINK, XTTS_CHECKPOINT_LINK], CHECKPOINTS_OUT_PATH, progress_bar=True
    )

 > Downloading DVAE files!


  0%|          | 0.00/1.07k [00:00<?, ?iB/s]
100%|██████████| 1.07k/1.07k [00:00<00:00, 3.50kiB/s]

  1%|          | 1.60M/211M [00:00<00:13, 16.0MiB/s]
  4%|▍         | 8.33M/211M [00:00<00:04, 46.1MiB/s]
  8%|▊         | 16.3M/211M [00:00<00:03, 61.5MiB/s]
 11%|█         | 23.3M/211M [00:00<00:02, 64.8MiB/s]
 15%|█▍        | 31.1M/211M [00:00<00:02, 69.6MiB/s]
 18%|█▊        | 38.9M/211M [00:00<00:02, 72.2MiB/s]
 22%|██▏       | 47.1M/211M [00:00<00:02, 75.6MiB/s]
 26%|██▌       | 54.9M/211M [00:00<00:02, 76.1MiB/s]
 30%|██▉       | 62.9M/211M [00:00<00:01, 77.3MiB/s]
 34%|███▎      | 70.7M/211M [00:01<00:01, 77.8MiB/s]
 37%|███▋      | 78.7M/211M [00:01<00:01, 78.3MiB/s]
 41%|████      | 86.6M/211M [00:01<00:01, 78.6MiB/s]
 45%|████▍     | 94.6M/211M [00:01<00:01, 79.0MiB/s]
 49%|████▊     | 103M/211M [00:01<00:01, 76.7MiB/s] 
 52%|█████▏    | 110M/211M [00:01<00:01, 77.0MiB/s]
 56%|█████▌    | 118M/211M [00:01<00:01, 69.4MiB/s]
 60%|█████▉    | 126M/211M [00:01<00:01, 71.3MiB/s]
 6

 > Downloading XTTS v2.0 files!


100%|██████████| 211M/211M [00:03<00:00, 65.5MiB/s]
 62%|██████▏   | 222k/361k [00:00<00:00, 2.21MiB/s]
100%|██████████| 361k/361k [00:00<00:00, 811kiB/s] 

  0%|          | 897k/1.87G [00:00<03:29, 8.90MiB/s]
  0%|          | 4.73M/1.87G [00:00<01:11, 26.1MiB/s]
  1%|          | 10.7M/1.87G [00:00<00:44, 41.4MiB/s]
  1%|          | 16.4M/1.87G [00:00<00:38, 47.6MiB/s]
  1%|          | 21.2M/1.87G [00:00<00:43, 42.6MiB/s]
  1%|▏         | 27.2M/1.87G [00:00<00:38, 48.2MiB/s]
  2%|▏         | 32.1M/1.87G [00:00<00:37, 48.4MiB/s]
  2%|▏         | 37.8M/1.87G [00:00<00:36, 50.8MiB/s]
  2%|▏         | 43.6M/1.87G [00:00<00:34, 53.0MiB/s]
  3%|▎         | 49.6M/1.87G [00:01<00:33, 55.0MiB/s]
  3%|▎         | 55.1M/1.87G [00:01<00:33, 54.5MiB/s]
  3%|▎         | 61.5M/1.87G [00:01<00:31, 57.4MiB/s]
  4%|▎         | 67.9M/1.87G [00:01<00:30, 59.2MiB/s]
  4%|▍         | 74.1M/1.87G [00:01<00:29, 60.0MiB/s]
  4%|▍         | 80.5M/1.87G [00:01<00:29, 61.4MiB/s]
  5%|▍         | 86.7M/1.87G [00:0

In [ ]:
training_dir = "/content/gdrive/MyDrive/data/dataset22050"

### Batch Size ###

* BATCH_SIZE is the amount of items being loaded into VRAM/memory at once.

* GRAD_ACCUM_STEPS is the amount of times we perform a forward pass with BATCH_SIZE amount of items before updating the parameters according to the SGD algorithm.

So a BATCH_SIZE of 2 and GRAD_ACCUM_STEPS of 32 would give an 'effective batch size' of 64 (i.e., 64 items considered per optimisation step).

The creators of XTTS-v2 recommend an effective batch size of 252 for proper training. (I found that reducing this to 126 gave me similar performance and faster fine-tuning, but we should probably listen to them.)

Your BATCH_SIZE will depend on how big your dataset items are and how much VRAM you have available. Make it as large as possible while ensuring that everything fits in VRAM and then make BATCH_SIZE*GRAD_ACCUM_STEPS==252.

In [ ]:

OPTIMIZER_WD_ONLY_ON_WEIGHTS = True
START_WITH_EVAL = True
BATCH_SIZE = 1
GRAD_ACUMM_STEPS = 1
LANGUAGE = "ar"

### Dataset Config

In [ ]:

model_args = GPTArgs(
    max_conditioning_length=242524,#the audio you will use for conditioning latents should be less than this
    min_conditioning_length=66168,#and more than this
    debug_loading_failures=True,#this will print output to console and help you find problems in your ds
    max_wav_length=530000,#set this to >= the longest audio in your dataset
    max_text_length=170,
    mel_norm_file=MEL_NORM_FILE,
    dvae_checkpoint=DVAE_CHECKPOINT,
    xtts_checkpoint=XTTS_CHECKPOINT,
    tokenizer_file=TOKENIZER_FILE,
    gpt_num_audio_tokens=1026,
    gpt_start_audio_token=1024,
    gpt_stop_audio_token=1025,
    gpt_use_masking_gt_prompt_approach=True,
    gpt_use_perceiver_resampler=True,
)

### Audio Config ###

The coqui TTS docs mention inspecting your data with the CheckSpectrograms.ipynb notebook to help decide on audio parameters. I think this is irrelevant for XTTS-v2, because it doesn't use the same audio config as some of the older coqui models and doesn't have the same parameters.

The default is 22050 for input and 24000 for output.

**the only reason my sample rate is lower is the dataset I chose has a sample rate of 16000. Assuming you are using your own data, sample_rate and dvae_sample_rate should match your data which should probably be 22050.**

In [ ]:
audio_config = XttsAudioConfig(sample_rate=22050, dvae_sample_rate=22050, output_sample_rate=24000)

### Speaker Reference ###


In [ ]:
SPEAKER_REFERENCE = "/content/gdrive/MyDrive/data/dataset22050/audio004-chunk65.wav"

### Trainer Config

In [ ]:
config = GPTTrainerConfig(
    run_eval=False,
    epochs = 15, # assuming you want to end training manually w/ keyboard interrupt
    output_path=OUT_PATH,
    model_args=model_args,
    run_name=RUN_NAME,
    project_name=PROJECT_NAME,
    run_description="""
        GPT XTTS training
        """,
    dashboard_logger=DASHBOARD_LOGGER,
    wandb_entity=None,
    logger_uri=LOGGER_URI,
    audio=audio_config,
    batch_size=BATCH_SIZE,
    batch_group_size=48,
    eval_batch_size=BATCH_SIZE,
    num_loader_workers=8, #consider decreasing if your jupyter env is crashing or similar
    eval_split_max_size=256,
    print_step=50,
    plot_step=100,
    log_model_step=1000,
    save_step=9999999999, #ALREADY SAVES EVERY EPOCHMaking this high on kaggle because Output dir is limited in size. I changed this to be size of training set/2 so I would effectively have a checkpoint every half epoch
    save_n_checkpoints=2,#if you want to store multiple checkpoint rather than just 1, increase this
    save_checkpoints=False,# Making this False on kaggle because Output dir is limited
    print_eval=False,
    optimizer="AdamW",
    optimizer_wd_only_on_weights=OPTIMIZER_WD_ONLY_ON_WEIGHTS,
    optimizer_params={"betas": [0.9, 0.96], "eps": 1e-8, "weight_decay": 1e-2},
    lr=5e-06,
    lr_scheduler="MultiStepLR",
    lr_scheduler_params={"milestones": [50000 * 18, 150000 * 18, 300000 * 18], "gamma": 0.5, "last_epoch": -1},
    test_sentences=[
        {
            "text": "في رفاهيه على جال هذيك هذاك الشي اللي بدا يصرى معها عجبها بزاف وشافت حياتها عمرها ما كانت تستنا ولا تتوقع انه يجي نهار تعيشها في حياتها",
            "speaker_wav": SPEAKER_REFERENCE,
            "language": LANGUAGE,
        },
    ],
)

# model = GPTTrainer.init_from_config(config)

### Load Dataset

In [ ]:
import torch
from TTS.tts.configs.xtts_config import XttsConfig
from TTS.tts.models.xtts import XttsAudioConfig, XttsArgs
from TTS.config.shared_configs import BaseDatasetConfig

torch.serialization.add_safe_globals([
    XttsConfig,
    XttsAudioConfig,
    BaseDatasetConfig,
    XttsArgs,  # ✅ New required class
])


In [ ]:
model = GPTTrainer.init_from_config(config)


>> DVAE weights restored from: /content/gdrive/MyDrive/models/EGTTS_v2.0_original_model_files/dvae.pth


In [ ]:
dataset_config = BaseDatasetConfig(
    formatter="ljspeech", meta_file_train="metadata.csv", language=LANGUAGE, path=training_dir
)
train_samples, eval_samples = load_tts_samples(dataset_config, eval_split=True, eval_split_size=0)

 | > Found 6910 files in /content/gdrive/.shortcut-targets-by-id/1-3wyUL1sMbtzmrrsYzXSSaj87srTsvTx/dataset22050


### Train! ###

**Note on warnings:**

The trainer will print out warnings if it encounters items in your dataset where the text exceeds 166 chars or the length of your audio exceeds max_wav_length (It will also have problems if you have data items that are < ~0.2s long, which you won't want anyway). You should remove these from your dataset or re-think how you're creating your dataset.




In [ ]:
import os

def clean_and_rename(samples):
    for sample in samples:
        original_path = sample['audio_file']

        # Fix the patha
        if original_path.endswith('.wav.wav'):
            fixed_path = original_path.replace('/wavs/', '/').replace('.wav.wav', '.wav')

            # Check if the incorrect file exists (for renaming)
            if os.path.exists(original_path):
                os.rename(original_path, fixed_path)
                print(f"Renamed: {original_path} -> {fixed_path}")
            elif os.path.exists(fixed_path):
                print(f"File already correct: {fixed_path}")
            else:
                print(f"File not found: {original_path} or {fixed_path}")

            # Update the sample dict
            sample['audio_file'] = fixed_path
            sample['audio_unique_name'] = '#' + fixed_path.replace(sample['root_path'], '').lstrip('/')
    return samples

# Apply the function
train_samples = clean_and_rename(train_samples)
eval_samples = clean_and_rename(eval_samples)


Streaming output truncated to the last 5000 lines.
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio002-chunk119.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio013-chunk70.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio008-chunk25.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio003-chunk232.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio021-chunk24.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio017-chunk11.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio003-chunk436.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio015-chunk83.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio021-chunk360.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio003-chunk1098.wav
File already correct: /content/gdrive/MyDrive/data/dataset22050/audio003-chunk132.wav
File al

In [ ]:
train_samples

[{'text': 'الماستر الاول جاني النقاط تاعي ما كانتش عاليه بزاف وعجبت يعني العام بصح المعدل تاعي ما كانش عالي بزاف باش لقيت باش نلقى\n',
  'audio_file': '/content/gdrive/MyDrive/data/dataset22050/audio018-chunk32.wav',
  'speaker_name': 'ljspeech',
  'root_path': '/content/gdrive/MyDrive/data/dataset22050',
  'language': 'ar',
  'audio_unique_name': '#audio018-chunk32.wav'},
 {'text': 'عملاء المشروع رقم 10 واللي هو جلب عقود العمل للشركات اون فيت هذ سي سامبل تروح تحكم شركه راهي تخدم\n',
  'audio_file': '/content/gdrive/MyDrive/data/dataset22050/audio022-chunk122.wav',
  'speaker_name': 'ljspeech',
  'root_path': '/content/gdrive/MyDrive/data/dataset22050',
  'language': 'ar',
  'audio_unique_name': '#audio022-chunk122.wav'},
 {'text': 'صح انا كنت دائما نعتبر انه الدراسه كانت تحدي كنت\n',
  'audio_file': '/content/gdrive/MyDrive/data/dataset22050/audio013-chunk217.wav',
  'speaker_name': 'ljspeech',
  'root_path': '/content/gdrive/MyDrive/data/dataset22050',
  'language': 'ar',
  'audio_un

In [ ]:
import torchaudio
for sample in train_samples:
    try:
        waveform, sr = torchaudio.load(sample["audio_file"])
        if waveform.abs().max() == 0:
            print("Silent file:", sample["audio_file"])
    except Exception as e:
        print("Error loading:", sample["audio_file"], e)


In [ ]:
trainer = Trainer(
    TrainerArgs(
        restore_path=None,
        skip_train_epoch=False,
        start_with_eval=START_WITH_EVAL,
        grad_accum_steps=GRAD_ACUMM_STEPS,
    ),
    config,
    output_path=OUT_PATH,
    model=model,
    train_samples=train_samples,
    eval_samples=eval_samples,
)


 > Training Environment:
 | > Backend: Torch
 | > Mixed precision: False
 | > Precision: float32
 | > Current device: 0
 | > Num. of GPUs: 1
 | > Num. of CPUs: 2
 | > Num. of Torch Threads: 1
 | > Torch seed: 1
 | > Torch CUDNN: True
 | > Torch CUDNN deterministic: False
 | > Torch CUDNN benchmark: False
 | > Torch TF32 MatMul: False
wandb: Currently logged in as: ba-alasmar (ba-alasmar-nit) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin



 > Model has 518442047 parameters


========Ignoring artifact: colab_kernel_launcher.py /usr/local/lib/python3.11/dist-packages/colab_kernel_launcher.py========


In [ ]:
trainer.fit()


 > EPOCH: 0/15
 --> /content/gdrive/MyDrive/models/finetuned_egtts_v2-May-17-2025_07+27AM-0000000
/usr/local/lib/python3.11/dist-packages/torch/utils/data/dataloader.py:624: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(

 > TRAINING (2025-05-17 07:27:43) 


 > Sampling by language: dict_keys(['ar'])



   --> TIME: 2025-05-17 07:27:51 -- STEP: 0/6841 -- GLOBAL_STEP: 0
     | > loss_text_ce: 0.038067929446697235  (0.038067929446697235)
     | > loss_mel_ce: 6.255189418792725  (6.255189418792725)
     | > loss: 6.293257236480713  (6.293257236480713)
     | > grad_norm: 0  (0)
     | > current_lr: 5e-06 
     | > step_time: 0.9837  (0.983717679977417)
     | > loader_time: 7.4155  (7.415520191192627)



========Ignoring artifact: checkpoint /content/gdrive/MyDrive/models/finetuned_egtts_v2-May-17-2025_07+27AM-0000000========



   --> TIME: 2025-05-17 07:28:13 -- STEP: 50/6841 -- GLOBAL_STEP: 50
     | > loss_text_ce: 0.039510346949100494  (0.03794555708765983)
     | > loss_mel_ce: 4.370471477508545  (4.82191523551941)
     | > loss: 4.409981727600098  (4.85986078262329)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.165  (0.17918388366699214)
     | > loader_time: 0.0172  (0.020978631973266607)


   --> TIME: 2025-05-17 07:28:35 -- STEP: 100/6841 -- GLOBAL_STEP: 100
     | > loss_text_ce: 0.03626270592212677  (0.03775879783555866)
     | > loss_mel_ce: 5.54624605178833  (4.69530585050583)
     | > loss: 5.582508563995361  (4.733064644336697)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1497  (0.17944274187088008)
     | > loader_time: 0.0166  (0.019799559116363528)


   --> TIME: 2025-05-17 07:28:56 -- STEP: 150/6841 -- GLOBAL_STEP: 150
     | > loss_text_ce: 0.035855840891599655  (0.037250470916430155)
     | > loss_mel_ce: 4.04953527450561

========Ignoring artifact: checkpoint /content/gdrive/MyDrive/models/finetuned_egtts_v2-May-17-2025_07+27AM-0000000========



   --> TIME: 2025-05-17 07:35:40 -- STEP: 1050/6841 -- GLOBAL_STEP: 1050
     | > loss_text_ce: 0.03552448749542236  (0.03502988612367993)
     | > loss_mel_ce: 4.253939628601074  (nan)
     | > loss: 4.289463996887207  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.149  (0.18867802869705932)
     | > loader_time: 0.0073  (0.013779914946783134)


   --> TIME: 2025-05-17 07:36:03 -- STEP: 1100/6841 -- GLOBAL_STEP: 1100
     | > loss_text_ce: 0.032954540103673935  (0.034960608981888394)
     | > loss_mel_ce: 4.279047966003418  (nan)
     | > loss: 4.312002658843994  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.2206  (0.1889902279593729)
     | > loader_time: 0.0098  (0.013679254705255663)


   --> TIME: 2025-05-17 07:36:26 -- STEP: 1150/6841 -- GLOBAL_STEP: 1150
     | > loss_text_ce: 0.03408007696270943  (0.03490278997337039)
     | > loss_mel_ce: 4.1787614822387695  (nan)
     | > loss: 4.212841510772705  (n

========Ignoring artifact: checkpoint /content/gdrive/MyDrive/models/finetuned_egtts_v2-May-17-2025_07+27AM-0000000========



   --> TIME: 2025-05-17 07:43:10 -- STEP: 2050/6841 -- GLOBAL_STEP: 2050
     | > loss_text_ce: 0.036594681441783905  (0.034346526241520514)
     | > loss_mel_ce: 4.314409255981445  (nan)
     | > loss: 4.351004123687744  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1739  (0.18958829437814118)
     | > loader_time: 0.0086  (0.012704580818734507)


   --> TIME: 2025-05-17 07:43:33 -- STEP: 2100/6841 -- GLOBAL_STEP: 2100
     | > loss_text_ce: 0.028479039669036865  (0.034320755614233825)
     | > loss_mel_ce: 3.3435168266296387  (nan)
     | > loss: 3.3719959259033203  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1643  (0.18970240445364106)
     | > loader_time: 0.0091  (0.012656230245317724)


   --> TIME: 2025-05-17 07:43:55 -- STEP: 2150/6841 -- GLOBAL_STEP: 2150
     | > loss_text_ce: 0.03398380056023598  (0.034306164564954646)
     | > loss_mel_ce: 4.29138708114624  (nan)
     | > loss: 4.32537078857421

========Ignoring artifact: checkpoint /content/gdrive/MyDrive/models/finetuned_egtts_v2-May-17-2025_07+27AM-0000000========



   --> TIME: 2025-05-17 07:50:33 -- STEP: 3050/6841 -- GLOBAL_STEP: 3050
     | > loss_text_ce: 0.026761429384350777  (0.03387641503613018)
     | > loss_mel_ce: 3.7062127590179443  (nan)
     | > loss: 3.7329742908477783  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1936  (0.1885805589644639)
     | > loader_time: 0.0102  (0.012234595173695052)


   --> TIME: 2025-05-17 07:50:55 -- STEP: 3100/6841 -- GLOBAL_STEP: 3100
     | > loss_text_ce: 0.034083493053913116  (0.033853383584008155)
     | > loss_mel_ce: 3.6821372509002686  (nan)
     | > loss: 3.7162208557128906  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1442  (0.1885121333214548)
     | > loader_time: 0.0084  (0.012208865227237829)


   --> TIME: 2025-05-17 07:51:18 -- STEP: 3150/6841 -- GLOBAL_STEP: 3150
     | > loss_text_ce: 0.03322644904255867  (0.033841812890318686)
     | > loss_mel_ce: 4.794926643371582  (nan)
     | > loss: 4.82815313339233

========Ignoring artifact: checkpoint /content/gdrive/MyDrive/models/finetuned_egtts_v2-May-17-2025_07+27AM-0000000========



   --> TIME: 2025-05-17 07:58:02 -- STEP: 4050/6841 -- GLOBAL_STEP: 4050
     | > loss_text_ce: 0.03321122005581856  (0.03356532114523433)
     | > loss_mel_ce: 4.120960235595703  (nan)
     | > loss: 4.154171466827393  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1443  (0.18908588921582287)
     | > loader_time: 0.0084  (0.012039103213651684)


   --> TIME: 2025-05-17 07:58:24 -- STEP: 4100/6841 -- GLOBAL_STEP: 4100
     | > loss_text_ce: 0.03420577943325043  (0.03355824916010227)
     | > loss_mel_ce: 4.133420467376709  (nan)
     | > loss: 4.16762638092041  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.2191  (0.18908753127586578)
     | > loader_time: 0.0314  (0.012042781667011544)


   --> TIME: 2025-05-17 07:58:46 -- STEP: 4150/6841 -- GLOBAL_STEP: 4150
     | > loss_text_ce: 0.030863381922245026  (0.033543715092342344)
     | > loss_mel_ce: 3.342435598373413  (nan)
     | > loss: 3.3732988834381104  (

error loading /content/gdrive/MyDrive/data/dataset22050/audio018-chunk1.wav (<class 'AssertionError'>, AssertionError('UNK token found in sociét \u200fsolution solaire\n -> [ar]sociét solution solaire'), <traceback object at 0x7d4d1580c9c0>)



   --> TIME: 2025-05-17 08:01:23 -- STEP: 4500/6841 -- GLOBAL_STEP: 4500
     | > loss_text_ce: 0.030558457598090172  (0.0334282848702537)
     | > loss_mel_ce: 3.3503623008728027  (nan)
     | > loss: 3.380920648574829  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1701  (0.18912683147854306)
     | > loader_time: 0.0086  (0.011982165760464139)


   --> TIME: 2025-05-17 08:01:44 -- STEP: 4550/6841 -- GLOBAL_STEP: 4550
     | > loss_text_ce: 0.03373276814818382  (0.03341199283784896)
     | > loss_mel_ce: 3.256610870361328  (nan)
     | > loss: 3.2903435230255127  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1588  (0.1890603565383747)
     | > loader_time: 0.0092  (0.011976811283237332)


   --> TIME: 2025-05-17 08:02:07 -- STEP: 4600/6841 -- GLOBAL_STEP: 4600
     | > loss_text_ce: 0.03292711079120636  (0.0334018612551787)
     | > loss_mel_ce: 5.251677513122559  (nan)
     | > loss: 5.284604549407959  (na

error loading /content/gdrive/MyDrive/data/dataset22050/audio015-chunk188.wav (<class 'AssertionError'>, AssertionError('UNK token found in \u200fendعمين وكيجي تريكروتي واحد وشنو هما زعما المعايير اللي تشوفهم وكيفاش تتعامل مع\n -> [ar]endعمين وكيجي تريكروتي واحد وشنو هما زعما المعايير اللي تشوفهم وكيفاش تتعامل مع'), <traceback object at 0x7d4d159b3c40>)



   --> TIME: 2025-05-17 08:02:31 -- STEP: 4650/6841 -- GLOBAL_STEP: 4650
     | > loss_text_ce: 0.035876695066690445  (0.03339293264934138)
     | > loss_mel_ce: 3.646162271499634  (nan)
     | > loss: 3.6820390224456787  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.3016  (0.18933531289459585)
     | > loader_time: 0.013  (0.011960969894163068)


   --> TIME: 2025-05-17 08:02:54 -- STEP: 4700/6841 -- GLOBAL_STEP: 4700
     | > loss_text_ce: 0.031106945127248764  (0.0333806658678867)
     | > loss_mel_ce: 4.064615249633789  (nan)
     | > loss: 4.095722198486328  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.21  (0.18946102421334474)
     | > loader_time: 0.0098  (0.011952768893952062)


   --> TIME: 2025-05-17 08:03:17 -- STEP: 4750/6841 -- GLOBAL_STEP: 4750
     | > loss_text_ce: 0.0353623665869236  (0.0333696488283183)
     | > loss_mel_ce: 4.942680358886719  (nan)
     | > loss: 4.9780426025390625  (nan)

========Ignoring artifact: checkpoint /content/gdrive/MyDrive/models/finetuned_egtts_v2-May-17-2025_07+27AM-0000000========



   --> TIME: 2025-05-17 08:05:32 -- STEP: 5050/6841 -- GLOBAL_STEP: 5050
     | > loss_text_ce: 0.03440452367067337  (0.03330845540444743)
     | > loss_mel_ce: 0.8833404779434204  (nan)
     | > loss: 0.9177449941635132  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.2614  (0.18989947323751954)
     | > loader_time: 0.0109  (0.011905660770907262)


   --> TIME: 2025-05-17 08:05:54 -- STEP: 5100/6841 -- GLOBAL_STEP: 5100
     | > loss_text_ce: 0.02960917353630066  (0.03330085200458472)
     | > loss_mel_ce: 4.913956642150879  (nan)
     | > loss: 4.943565845489502  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1891  (0.1898918646457151)
     | > loader_time: 0.0115  (0.011909314931607708)


   --> TIME: 2025-05-17 08:06:18 -- STEP: 5150/6841 -- GLOBAL_STEP: 5150
     | > loss_text_ce: 0.032879143953323364  (0.03329099250649941)
     | > loss_mel_ce: 4.731638431549072  (nan)
     | > loss: 4.764517784118652  (

error loading /content/gdrive/MyDrive/data/dataset22050/audio018-chunk1.wav (<class 'AssertionError'>, AssertionError('UNK token found in sociét \u200fsolution solaire\n -> [ar]sociét solution solaire'), <traceback object at 0x7d4d157bd840>)



   --> TIME: 2025-05-17 08:09:15 -- STEP: 5550/6841 -- GLOBAL_STEP: 5550
     | > loss_text_ce: 0.0387418270111084  (0.03321819150132369)
     | > loss_mel_ce: 3.615011215209961  (nan)
     | > loss: 3.6537530422210693  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.2647  (0.18977873183585542)
     | > loader_time: 0.0087  (0.011886677398337979)


   --> TIME: 2025-05-17 08:09:36 -- STEP: 5600/6841 -- GLOBAL_STEP: 5600
     | > loss_text_ce: 0.029210058972239494  (0.033206756102652946)
     | > loss_mel_ce: 3.489591360092163  (nan)
     | > loss: 3.51880145072937  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1543  (0.18966066654239422)
     | > loader_time: 0.0093  (0.011871521600655152)


   --> TIME: 2025-05-17 08:09:58 -- STEP: 5650/6841 -- GLOBAL_STEP: 5650
     | > loss_text_ce: 0.03015930764377117  (0.033197811276248046)
     | > loss_mel_ce: 3.438607931137085  (nan)
     | > loss: 3.4687671661376953  

========Ignoring artifact: checkpoint /content/gdrive/MyDrive/models/finetuned_egtts_v2-May-17-2025_07+27AM-0000000========



   --> TIME: 2025-05-17 08:12:59 -- STEP: 6050/6841 -- GLOBAL_STEP: 6050
     | > loss_text_ce: 0.02988494746387005  (0.0331337866550388)
     | > loss_mel_ce: 3.5177924633026123  (nan)
     | > loss: 3.547677516937256  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1917  (0.189886901437744)
     | > loader_time: 0.014  (0.011844144813285391)


   --> TIME: 2025-05-17 08:13:22 -- STEP: 6100/6841 -- GLOBAL_STEP: 6100
     | > loss_text_ce: 0.029798252508044243  (0.03312297447477693)
     | > loss_mel_ce: 3.45151948928833  (nan)
     | > loss: 3.4813177585601807  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.131  (0.18997769750532595)
     | > loader_time: 0.0081  (0.011843033735869365)


   --> TIME: 2025-05-17 08:13:45 -- STEP: 6150/6841 -- GLOBAL_STEP: 6150
     | > loss_text_ce: 0.022588102146983147  (0.03310943317304293)
     | > loss_mel_ce: 3.2135818004608154  (nan)
     | > loss: 3.2361698150634766  (na

error loading /content/gdrive/MyDrive/data/dataset22050/audio001-chunk416.wav (<class 'AssertionError'>, AssertionError('UNK token found in \u200ff g it was good f\n -> [ar]f g it was good f'), <traceback object at 0x7d4d157bd240>)



   --> TIME: 2025-05-17 08:17:31 -- STEP: 6650/6841 -- GLOBAL_STEP: 6650
     | > loss_text_ce: 0.03128884732723236  (0.033025847372871994)
     | > loss_mel_ce: 3.036968231201172  (nan)
     | > loss: 3.0682570934295654  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1488  (0.19037754216588565)
     | > loader_time: 0.01  (0.011827259458097305)


   --> TIME: 2025-05-17 08:17:52 -- STEP: 6700/6841 -- GLOBAL_STEP: 6700
     | > loss_text_ce: 0.028129301965236664  (0.03301809676464144)
     | > loss_mel_ce: 3.4550302028656006  (nan)
     | > loss: 3.4831595420837402  (nan)
     | > grad_norm: 0  (0.0)
     | > current_lr: 5e-06 
     | > step_time: 0.1336  (0.19032424428569772)
     | > loader_time: 0.008  (0.011820135828274403)


   --> TIME: 2025-05-17 08:18:14 -- STEP: 6750/6841 -- GLOBAL_STEP: 6750
     | > loss_text_ce: 0.0328197106719017  (0.0330097917925428)
     | > loss_mel_ce: 4.043896675109863  (nan)
     | > loss: 4.076716423034668  (na

 | > Synthesizing test sentences.
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/trainer/trainer.py", line 1833, in fit
    self._fit()
  File "/usr/local/lib/python3.11/dist-packages/trainer/trainer.py", line 1789, in _fit
    self.test_run()
  File "/usr/local/lib/python3.11/dist-packages/trainer/trainer.py", line 1698, in test_run
    test_outputs = self.model.test_run(self.training_assets)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/TTS/tts/layers/xtts/trainer/gpt_trainer.py", line 235, in test_run
    wav = self.xtts.synthesize(
          ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/TTS/tts/models/xtts.py", line 419, in synthesize
    return self.full_inference(text, speaker_wav, languag

TypeError: object of type 'NoneType' has no len()